contracts/swap/swap.go

```diff
@@ -0,0 +1,171 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package swap wraps the 'swap' Ethereum smart contract.
+// It is an abstraction layer to hide implementation details about the different
+// Swap contract iterations (Simple Swap, Soft Swap, etc.)
+package swap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/swarm/contracts/swap/contract"
+)
+
+var (
+	// ErrNotASwapContract is given when an address is verified not to have a SWAP contract based on its bytecode
+	ErrNotASwapContract = errors.New("not a swap contract")
+	// ErrTransactionReverted is given when the transaction that submits or cashes a cheque is reverted
+	ErrTransactionReverted = errors.New("Transaction reverted")
+)
+
+// Backend wraps all methods required for contract deployment.
+type Backend interface {
+	bind.ContractBackend
+	TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error)
+}
+
+// Deploy deploys an instance of the underlying contract and returns its `Contract` abstraction
+func Deploy(auth *bind.TransactOpts, backend bind.ContractBackend, owner common.Address, harddepositTimeout time.Duration) (common.Address, Contract, *types.Transaction, error) {
+	addr, tx, s, err := contract.DeploySimpleSwap(auth, backend, owner, big.NewInt(int64(harddepositTimeout)))
+	c := simpleContract{instance: s}
+	return addr, c, tx, err
+}
+
+// InstanceAt creates a new instance of a contract at a specific address.
+// It assumes that there is an existing contract instance at the given address, or an error is returned
+// This function is needed to communicate with remote Swap contracts (e.g. sending a cheque)
+func InstanceAt(address common.Address, backend bind.ContractBackend) (Contract, error) {
+	simple, err := contract.NewSimpleSwap(address, backend)
+	c := simpleContract{instance: simple}
+	return c, err
+}
+
+// Contract interface defines the methods exported from the underlying go-bindings for the smart contract
+type Contract interface {
+	// Submit a cheque to the beneficiary
+	SubmitChequeBeneficiary(opts *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Receipt, error)
```

Comment:
@Eknir @ralph-pichler your input would be appreciated here

Reply:
All contract interaction always uses `big.Int` as its the only go type that is suitable for evm numbers. For now the cheque structure in the go code uses `uint64` and it's those values that are converted to `big.Int` and used to call this function. Therefore this cannot overflow in practice.

In the future we will have to use `big.Int` (or something similar) in the cheque structure. Then we will have to verify the range (0 to 2^256-1) when we receive the cheque and reject it otherwise.

Reply:
related issue: #1581